### DISCUSSION TITLE
Markdown math support (Latex)

### DISCUSSION THREAD

**Discussion by Vagab** [zed-industries/zed#24224] [Feature Requests]
### Check for existing issues

- [X] Completed

### Describe the feature

Would be cool if markdown preview rendered math as well

**Comment by ferntheplant** [comment]
It seems like this would be very non-trivial to implement. Zed uses its own customer renderer so they'd have to make a LaTeX renderer that produces something like `MathElement`. I didn't dig super deep into that side of the equation but think that once that is done it would be relatively easy to regex match for `$` enclosed text and render `MathElement` as part of the [`render_markdown_text`](https://github.com/zed-industries/zed/blob/87457f9ae8d55a579e456f480e5813f89c105af5/crates/markdown_preview/src/markdown_renderer.rs#L337) sequence that currently calculates highlights and generates interactive text elements for hyperlinks.

**Comment by bbb651** [comment]
`pulldown-cmark` [added support for it](https://github.com/pulldown-cmark/pulldown-cmark/pull/734) so should be easy on the parsing front. ~~Rendering~~ Layout is a lot more complicated, the only potentially useful reference I've found is [typst's implementation](https://github.com/typst/typst/tree/main/crates/typst/src/math).

**Comment by bitfield** [comment]
Just to give an idea of what this might look like (screenshot from the VS Code Markdown Preview extension).

The markup:

```text
1.11 Neptunium. In the fall of 2002, a group of scientists at Los Alamos National Laboratory determined 
that the critical mass of neptunium-237 is about 60 kg. The critical mass of a fissionable material is the 
minimum amount that must be brought together to start a chain reaction. This element has a density of 
19.5 g/cm^3. What would be the radius of a sphere of this material that has a critical mass?

Volume of the critical sphere $$ V_{crit} = 60,000 / 19.5 = 3077cm^3 = \frac{4\pi r^3}{3} $$

So

$$ R_{crit} = \sqrt[3]{\frac{3 \times 3077}{4 \pi}} = 9cm \; ✅ $$
```

Rendering:

![Screenshot 2024-08-22 at 18 41 51](https://github.com/user-attachments/assets/b5fe1cb7-ee43-4759-8ffa-63cd606c2091)

**Comment by bersace** [comment]
codestral produce markdown math. Very hard to read in assistant panel:

![Image](https://github.com/user-attachments/assets/355fd6dc-dad0-4218-8030-83bb0a249dbb)

**Comment by probably-neb** [comment]
This is definitely a neat idea, but as @ferntheplant probably non-trivial to implement given our custom renderer.

I'd definitely be interested in seeing this in the markdown renderer, especially as we get more support for Jupyter notebooks that would benefit from this kind of thing.

In order to keep track of this i'm going to convert this to a discussion because we've decided to move feature/enhancement requests there so it will be more likely someone from the team will see it there when looking for new features to add

**Comment by 497e0bdf29873** [comment]
As a temporary workaround that should be quicker to implement, how about triggering (on save or another key) markdown compilation externally (with markdown-it, which has good support for all sorts of features) and then launching a browser? If feasible, inserting an anchor point at cursor location and directing the browser to that anchor.

This is a similar pipeline as in the TeX package so maybe it could be adapted.

**Comment by lnay** [comment]
@ferntheplant I played with something along these lines [this week](https://github.com/lnay/reMathML). The idea is to render MathML into a bitmap-like format currently already used by GPUI for SVGs, and then something like Temml (or something in rust) could generate the MathML. Zed already uses node for `prettier` formatting so there is a precedent for running a node process to process text.

*Edit: this was just a proof of concept and is much less developed compared to KenyC/ReX I talk about in my other response.*

**Reply by paperbenni** [reply]
Pretty much this idea exact is already implemented in snacks.nvim
https://github.com/folke/snacks.nvim/blob/main/docs/image.md
It works remarkably well, it compiles all latex asynchronously and caches the results of individual math blocks.
Zed could download tectonic like it does with language servers upon opening a markdown file with math blocks, insert the math blocks into a latex template, render that using tectonic and display the result in the markdown preview buffer.

**Reply by lnay** [reply]
For clarity, my suggestion didn't involve invoking a 'proper' LaTeX compiler like pdflatex or tectonic as is done in snack.nvim. It would see the rendering more tightly integrated into GPUI similarly to how SVGs are currently dealt with.

**Reply by 497e0bdf29873** [reply]
KaTeX is almost standard for rendering math in markdown, and generates some sort of HTML/MathML. It's also Node.js.

**Reply by lnay** [reply]
It's the HTML/MathML that's the issue here, Zed's markdown renderer does not involve a webview.

**Comment by lnay** [comment]
I found a project that could be a could fit for maths rendering in GPUI (https://github.com/KenyC/ReX), here's a minimal demo with gpui: https://github.com/lnay/latex-gpui-experiment

KenyC/ReX is very modular and has been used already to have live preview tools: https://github.com/KenyC/maths-preview

**Reply by michaelfortunato** [reply]
This seems like a good approach perhaps one day we can even render latex in lsp hover windows

**Comment by axeljeremy7** [comment]
so there is no way to preview math on Zed?

**Reply by dawsh2** [reply]
I don't see why you couldn't get it working with markdown preview + SVGs as a temporary working solution. I'm gonna try it later.

*Why has this not been mentioned?
https://github.com/lnay/zed-texpresso

**Reply by lnay** [reply]
The discussion is about latex in markdown to be previewed inside Zed. For other situations there are other solutions: 
- preview on save in an external window for a latex document: the latex extension 
- preview as you type in an external document for a latex (xelatex compliant): texpresso (requires more effort to set up)
- preview on saved markdown files with latex, in an external window: I'm sure something could be set up with tasks and pandoc or similar
- preview as you type of markdown including latex, inside Zed: this does not exist

ATM there is actually a PR for the last one with an adhoc attempt to represent equations in plain text with non-ascii characters.

**Reply by lnay** [reply]
Dawsh, seeing now what you did with mermaid diagrams in markdown, there is a project by KenyC which will output to create SVG from latex snippets. Maybe you could do something similar

**Reply by dawsh2** [reply]
Awesome I’ll check it out later, should be pretty similar to how the mermaid preview works.

**Reply by axeljeremy7** [reply]
im on mac and new on Zed and i dont see preview markdown math rendering at all, but can anyone explain , for instance to try this:

Let the Loss function be the average over $N$ data points:
$$ L(\theta) = \frac{1}{N} \sum_{i=1}^N \ell(f(x_i; \theta), y_i) $$

to see if render becuase on vscode is automatically

**Reply by lnay** [reply]
The short answer is there's no latex in markdown rendering in Zed. There was only some talk about what could be done about it.

**Comment by lucashmorais** [comment]
Thank you for everyone working on the issue, I just wanted to leave it here that the lack of this feature is the single reason why I dropped Zed after a few days of testing. I would go back the second it is implemented.

**Comment by auriium2** [comment]
Made an attempt with [this pr](https://github.com/zed-industries/zed/pull/44525), has functional latex support in the repl via svg rendering (latex -> mitex -> typst -> typst-svg) and could probably be extrapolated for markdown math support since svg rendering has been added. Probably can be superceded with better parser + gpui native render component at later time

**Comment by pnyda** [comment]
[Markdown Preview Language Server extension](https://github.com/zed-industries/extensions/pull/4190) lets you preview a markdown buffer in a browser tab 

https://github.com/user-attachments/assets/669bc099-aa68-4b89-ba84-32618ddb60c4

**Reply by revanthp** [reply]
@pnyda This does what I am looking for. thank you for the contribution. How did you get the local server running after installing the extension? Could you provide step by step instructions?

**Comment by lstwn** [comment]
I've encountered [RaTeX](https://ratex.lites.dev) (Rust-based, no WebView) which is probably a great fit for Zed's rendering pipeline, as it (apparently) can be used with a custom renderer! Might be worth a try :)

**Comment by MadeInShineA** [comment]
+1 For this killer feature

**Comment by bobbymannino** [comment]
https://github.com/zed-industries/zed/pull/57339 (not my work)

**Comment by Cyfine** [comment]
Any updates?

**Comment by Arcitec** [comment]
There are three competing RaTeX implementation pull requests:

- https://github.com/zed-industries/zed/pull/54668
- https://github.com/zed-industries/zed/pull/57339
- https://github.com/zed-industries/zed/pull/58593

But keep in mind that LaTeX support was already closed in 2024 as "not planned":

https://github.com/zed-industries/zed/issues/11126

And confirmed again in 2026 as "out of scope and not on our roadmap" because Zed is a code editor. It's not a LaTeX math formula renderer:

https://github.com/zed-industries/zed/pull/54668#issuecomment-4305857145

But it was reopened (without comment), so maybe there's a chance. If so, someone will have to take the time to figure out which of all the pull requests is the cleanest solution. And more importantly: Explain why Zed needs to render math formulas.

**Reply by MadeInShineA** [reply]
Why have a Markdown preview if not supporting those basic features lol (math's mermaid)

Makes the renderer useless if wanting to deploy the md on for example GitHub

**Reply by Cyfine** [reply]
Logic is simple: people write code to solve computational problems other than front end or backend, more or less involving math. People who use the editor taking notes involve math. Almost all modern Markdown editors support math rendering. There have been such feature requests since 2024 that means people need it. There is a FaceTime call built into the code editor, but math rendering in Markdown is deemed out of scope is confusing. The only thing keep me from using zed all the time is I have to switch to vscode to view math.

**Reply by akirakoyasu** [reply]
To add another perspective on why this would be useful in a code editor like Zed:

I'm a software engineer and a big fan of Zed, and I'm very grateful to everyone contributing to it.

These days, **I do almost all of my daily engineering work in Zed together with AI** — not only coding, but also software design, code reviews, technical research, and many other tasks. As more of my work happens this way, Markdown has also become a natural part of the workflow. Design notes, research results, implementation plans, explanations, and other technical artifacts are often written or generated in Markdown alongside the code.

In that context, **math rendering becomes surprisingly important. AI models naturally use LaTeX** when explaining algorithms, equations, complexity, statistics, machine learning concepts, and many other technical topics. When that happens, **being able to render the math directly in Zed's Markdown Preview would make it much easier** to read and review the content without leaving the editor.

For me, this isn't about turning Zed into a LaTeX editor. It's **simply about making Markdown fully readable within the engineering workflow that already happens inside Zed.**

I'd love to see math rendering supported in Markdown Preview.

**Reply by 497e0bdf29873** [reply]
I have moved most of my editing to Zed, but I still have to keep VSCode around for my note-taking. Because the markdown Preview in Zed is simply sub-par. Missing `$maths$` , missing `==highlights==`, etc. Does the maths hatred come from the Rust side? It's also painful to write documentation for computational Rust code, because rustdoc, similar to Zed, does not support maths (without very dirty hacks and lots of symbol escapes). Rust is great otherwise, but stuck in 80's gringo ASCII world. Same, apparently, with Zed.

**Reply by ethanwillis** [reply]
The reason? Because I am trained as a computer scientist.

Also I think the reason we're only having this conversation at all for this long is... If Zed was *only* about editing we wouldn't have a markdown *renderer* at all. Only a markdown *editor*!

This theme reaches further than just the markdown editor/markdown renderer. We have a code editor coupled with a rendering process for a terminal which can display software running in a terminal. If we didn't care about rendering things aside from their raw format I wouldn't be able to encode colored output in Zed's terminal!

**Reply by swfsql** [reply]
I've been using [zettlr](https://zettlr.com/) to preview the markdown with math support. Happy with with, the only problem is that it seems to not auto-reload if I change the files externally.  
note: the feature that changes a line from preview back to raw if the cursor is on it is pretty neat.

**Comment by adamgogacz** [comment]
+1

**Comment by emjomi** [comment]
+1

**Comment by gavmor** [comment]
I write and read a lot of ADRs in zed, and often those involve debating operational or performance tradeoffs, and a little bit of arithmetic helps get the point across. LaTeX is helpful for that.

They're co-located with the affected source code.
